File: CSharp/Specs/ThingTest.cs
#region

using System;
using NUnit.Framework;
using ThingModel.Builders;

#endregion

namespace ThingModel.Specs
{
    [TestFixture]
    class ThingTest
    {
        private Thing _thing;
        private ThingType _type;
        private Thing _otherThing;
        private Property _name;

        [SetUp]
        protected void SetUp()
        {
            _type = new ThingType("Person");
            _type.DefineProperty(PropertyType.Create<Property.String>("name"));

            _thing = new Thing("200787", _type);


            _otherThing = new Thing("875402");
            _thing.Connect(_otherThing);

            _name = new Property.String("name", "Pierre");
            _thing.SetProperty(_name);
        }

        [Test]
        public void CheckID()
        {
            Assert.That(_thing.ID, Is.EqualTo("200787"));
        }

        [Test]
        public void WrongID()
        {
            Assert.Throws<Exception>(() => new Thing(""));
        }

        [Test]
        public void Connect()
        {
            Assert.That(_thing.IsConnectedTo(_otherThing), Is.True);
			_thing.Connect(null);
            Assert.That(_thing.IsConnectedTo(null), Is.False);
			_thing.Disconnect(null);
        }

        [Test]
        public void Disconnect()
        {
            _thing.Disconnect(_otherThing);
            Assert.That(_thing.IsConnectedTo(_otherThing), Is.False);
        }

        [Test]
        public void CheckName()
        {
            Assert.That(_thing.GetProperty<Property.String>("name").Value, Is.EqualTo("Pierre"));
        }

        [Test]
        public void HasProperty()
        {
            Assert.That(_thing.HasProperty("name"), Is.True);
            Assert.That(_thing.HasProperty("lapin"), Is.False);
        }

        [Test]
        public void GetPropertyWithWrongType()
        {
            Assert.That(_thing.GetProperty<Property.Double>("name"), Is.Null);
        }

        [Test]
        public void GetNonexistentProperty()
        {
            Assert.That(_thing.GetProperty<Property.Boolean>("abcdef"), Is.Null);
            Assert.That(_thing.GetProperty<Property.String>("abcdef"), Is.Null);
        }

        [Test]
        public void CheckGetProperties()
        {
            
            var visited = false;
            foreach (var property in _thing.GetProperties())
            {
                visited = true;
                // Should not throw exceptions
                _thing.SetProperty(property);
            }

            Assert.That(visited, Is.True);
        }

        [Test]
        public void CheckObjet()
        {
            Assert.That(_type.Check(_thing), Is.True);

            _type.DefineProperty(PropertyType.Create<Property.Boolean>("beautiful"));
            Assert.That(_type.Check(_thing), Is.False);

            _thing.SetProperty(new Property.Location.Equatorial("beautiful", new Location.Equatorial()));
            Assert.That(_type.Check(_thing), Is.False);

            _type.DefineProperty(PropertyType.Create<Property.Location.Equatorial>("beautiful"));
            Assert.That(_type.Check(_thing), Is.True);
        }

        [Test]
        public void ConnectThingToItself()
        {
            Assert.Throws<Exception>(() => _thing.Connect(_thing));

            Assert.Throws<Exception>(() => _thing.Connect(new Thing(_thing.ID)));
        }

        [Test]
        public void CheckEnumerator()
        {
            Assert.That(_thing.ConnectedThings, Is.Not.Empty);

            // The connectedThing should be a copy of Connections if we delete or add things
            // during this loop
            foreach (var connectedThing in _thing.ConnectedThings)
            {
                _thing.Disconnect(connectedThing);
            }

            Assert.That(_thing.ConnectedThings, Is.Empty);
        }

        [Test]
        public void CompareSameThing()
        {
            Assert.That(_thing.Compare(_thing), Is.True);
        }

        [Test]
        public void CompareWithWrongType()
        {
            var newThing = new Thing(_thing.ID); // default type

            newThing.SetProperty(new Property.String("name", "Pierre"));
            newThing.Connect(_otherThing);

            Assert.That(_thing.Compare(newThing), Is.False);
        }

        [Test]
        public void CompareNotConnected()
        {
            var newThing = new Thing(_thing.ID, _type);

            newThing.SetProperty(new Property.String("name", "Pierre"));

            Assert.That(_thing.Compare(newThing), Is.False);
        }

        [Test]
        public void CompareWithDifferentID()
        {
            var newThing = new Thing("banana", _type);
            newThing.SetProperty(new Property.String("name", "Pierre"));
            newThing.Connect(_otherThing);

            Assert.That(_thing.Compare(newThing, false), Is.True);
            Assert.That(newThing.Compare(_thing, false), Is.True);

            // Different ID
            Assert.That(_thing.Compare(newThing), Is.False);
        }

        [Test]
        public void CompareProperties()
        {
            var newThing = new Thing(_thing.ID, _type);
            newThing.SetProperty(new Property.String("name", "Pierre"));
            newThing.SetProperty(new Property.String("surname", "Lapinou"));
            newThing.Connect(_otherThing);

            Assert.That(_thing.Compare(newThing), Is.False);
            Assert.That(newThing.Compare(_thing), Is.False);

            _thing.SetProperty(new Property.String("surname", "Lapinou"));
            Assert.That(_thing.Compare(newThing), Is.True);
            Assert.That(newThing.Compare(_thing), Is.True);

            _thing.SetProperty(new Property.String("surname", "Roger"));
            Assert.That(_thing.Compare(newThing), Is.False);
            Assert.That(newThing.Compare(_thing), Is.False);

            _thing.SetProperty(new Property.String("surname", "Lapinou"));
            _thing.SetProperty(new Property.String("color", "white"));
            newThing.SetProperty(new Property.String("meal", "nothing"));
            Assert.That(_thing.Compare(newThing), Is.False);
            Assert.That(newThing.Compare(_thing), Is.False);
        }

        [Test]
        public void DeepCompairison()
        {
            var aThingForTheRoad = new Thing("rabbit");
            aThingForTheRoad.SetProperty(new Property.Double("speed", 12));
            _thing.Connect(aThingForTheRoad);

            var newThing = new Thing(_thing.ID, _type);
            newThing.SetProperty(new Property.String("name", "Pierre"));
            newThing.Connect(_otherThing);
            newThing.Connect(aThingForTheRoad);

            Assert.That(_thing.Compare(_thing, true, true), Is.True);
            Assert.That(_thing.Compare(newThing, true, true), Is.True);
            Assert.That(newThing.Compare(_thing, true, true), Is.True);

            newThing.Disconnect(_otherThing);
            var newOtherThing = new Thing(_otherThing.ID);
			newThing.Connect(newOtherThing);

            Assert.That(_thing.Compare(newThing, true, true), Is.True);
            Assert.That(newThing.Compare(_thing, true, true), Is.True);

            newOtherThing.SetProperty(new Property.String("name", "Alain"));

            Assert.That(_thing.Compare(newThing, true, true), Is.False);
            Assert.That(newThing.Compare(_thing, true, true), Is.False);
        }

        [Test]
        public void InfiniteLoopDeepCompairaison()
        {
            var newThing = new Thing(_thing.ID, _type);
            newThing.SetProperty(new Property.String("name", "Pierre"));
            newThing.Connect(_otherThing);

            // Magic
            _otherThing.Connect(_thing);

            Assert.That(_thing.Compare(newThing, true, true), Is.True);
            Assert.That(newThing.Compare(_thing, true, true), Is.True);

            newThing.Disconnect(_otherThing);

            var newOtherThing = new Thing(_otherThing.ID);
            newThing.Connect(newOtherThing);

            Assert.That(_thing.Compare(newThing, true, true), Is.False);
            Assert.That(newThing.Compare(_thing, true, true), Is.False);

            newOtherThing.Connect(newThing);

            Assert.That(_thing.Compare(newThing, true, true), Is.True);
            Assert.That(newThing.Compare(_thing, true, true), Is.True);
        }

        [Test]
        public void CompareSomethingNull()
        {
            Assert.That(_thing.Compare(null), Is.False);
        }

	    [Test]
	    public void TestNewSyntax()
	    {
		    var thing = new Thing("canard");
		    thing.String("lapin", "canard")
			    .LocationEquatorial("location", new Location.Equatorial())
			    .Boolean("moving", true)
			    .DateTime("creation", DateTime.Now)
			    .Int("speed", 132)
			    .Double("vertical_speed", 145.66);

		    thing.ContainingA.String("name", "Canard2")
			    .AndA.Boolean("second", false);

			Assert.That(thing.String("lapin"), Is.EqualTo("canard"));
			Assert.That(thing.LocationEquatorial("location"), Is.EqualTo(new Location.Equatorial()));
		    Assert.That(thing.Boolean("moving"), Is.EqualTo(true));
		    Assert.That(thing.DateTime("creation"), Is.LessThanOrEqualTo(DateTime.Now));
		    Assert.That(thing.Int("speed"), Is.EqualTo(132));
		    Assert.That(thing.Double("vertical_speed"), Is.EqualTo(145.66));
		    Assert.That(thing.String("name"), Is.EqualTo("Canard2"));
		    Assert.That(thing.Boolean("second"), Is.EqualTo(false));
	    }

        [Test]
        public void TestClone()
        {
            Thing pond = BuildANewThing.WithoutType().IdentifiedBy("pond");
            Thing duck = BuildANewThing.WithoutType().IdentifiedBy("canard")
                .ContainingAn.Int("age", 12);

            duck.Connect(pond);

            var duck2 = duck.Clone();
            duck2.Int("age", 8);

            Assert.That(duck.Int("age"),Is.EqualTo(12));
            Assert.That(duck2.IsConnectedTo(pond), Is.True);
            Assert.That(duck2.ConnectedThings[0], Is.Null);

            duck2 = duck.Clone(true);
            Assert.That(duck2.Int("age"),Is.EqualTo(12));
            Assert.That(duck2.IsConnectedTo(pond), Is.True);
            Assert.That(duck2.ConnectedThings[0], Is.Not.Null);
            Assert.That(duck2.ConnectedThings[0], Is.EqualTo(pond));
        }
    }
}

File: TypeScript/Specs/ThingTest.ts
/// <reference path="../bower_components/DefinitelyTyped/mocha/mocha.d.ts" />
/// <reference path="../bower_components/DefinitelyTyped/should/should.d.ts" />

import should = require('should');

describe("Thing test", ()=> {
	var thing : ThingModel.Thing;
	var type : ThingModel.ThingType;
	var otherThing : ThingModel.Thing;
	var name : ThingModel.Property;

	beforeEach(() => {
		type = new ThingModel.ThingType("Person");
		type.DefineProperty(
			new ThingModel.PropertyType("name", ThingModel.Type.String));

		thing = new ThingModel.Thing("200787", type);

		otherThing = new ThingModel.Thing("875402");
		thing.Connect(otherThing);

		name = new ThingModel.Property.String("name", "Pierre");
		thing.SetProperty(name);

	});

// ReSharper disable WrongExpressionStatement
	describe("testWrongID", () => {
		it("should fail with a null", ()=> {
			(() => { thing = new ThingModel.Thing(null); }).should.throw(/null/);
		});

		it("should fail with an empty string", () => {
			(() => { thing = new ThingModel.Thing(""); }).should.throw(/empty/);
		});
	});

	describe("Test set properties", ()=> {
		it("should contains the saved value", ()=> {
			thing.SetProperty(new ThingModel.Property.Int("speed", 12));
			thing.SetProperty(new ThingModel.Property.Double("speed", 12.5));

			thing.GetProperty<ThingModel.Property.Double>("speed",
				ThingModel.Type.Double).Value.should.be.equal(12.5);
		});
	});

	describe("Test get property", ()=> {
		it("is not null", () => {
			(thing.GetProperty<ThingModel.Property>("name", ThingModel.Type.String) !== null).should.be.ok;
			(thing.GetProperty<ThingModel.Property>("name") !== null).should.be.ok;
			(thing.GetProperty<ThingModel.Property>("name", null) !== null).should.be.ok;

			(thing.GetProperty<ThingModel.Property>("name", ThingModel.Type.Double) === null).should.be.ok;
			(thing.GetProperty<ThingModel.Property>("wrong key", ThingModel.Type.String) === null).should.be.ok;
		});
	});

	describe("Test has property", ()=> {
		it("just few checks", ()=> {
			thing.HasProperty("name").should.be.true;
			thing.HasProperty("name2").should.be.false;
		});
	});

	describe("Test connections", ()=> {
		it("should be connected to connected things", ()=> {
			thing.IsConnectedTo(otherThing).should.be.true;
			otherThing.IsConnectedTo(thing).should.be.false;
		});

		it("should handle null values", ()=> {
			thing.Connect(null);
			thing.IsConnectedTo(null).should.be.false;
			thing.Disconnect(null);
		});

		it("should not be connected to itself", ()=> {
			(()=> {
				thing.Connect(thing);
			}).should.throw(/itself/);

			(() => {
				thing.Connect(new ThingModel.Thing(thing.ID));
			}).should.throw(/itself/);
		});

		it("should disconnect a thing properly", ()=> {
			thing.Disconnect(otherThing);
			thing.IsConnectedTo(otherThing).should.be.false;
		});

		it("should disconnect all things if needed", ()=> {
			thing.DisconnectAll();
			thing.IsConnectedTo(otherThing).should.be.false;
		});

		it("should create a list of connected things", () => {
			_.each(thing.ConnectedThings, (t) => {
				thing.Disconnect(t);
			});

			thing.IsConnectedTo(otherThing).should.be.false;
		});

		it("should count connected things correctly", ()=> {
			thing.ConnectedThingsCount.should.be.equal(1);
			thing.Disconnect(otherThing);
			thing.ConnectedThingsCount.should.be.equal(0);
		});
	});

	describe("Test properties", ()=> {
		it("should contains the name property", () => {
			_.each(thing.Properties, (p: ThingModel.Property) => {
				p.Key.should.be.equal("name");
			});
		});
	});

	describe("Test comparisons", () => {
		it("should handle nulls", () => {
			thing.Compare(null).should.be.false;
		});

		it("should return true when it's the same thing", () => {
			thing.Compare(thing).should.be.true;
		});

		it("should be different when the type is different", ()=> {
			var newThing = new ThingModel.Thing(thing.ID); // default type

			newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));
			newThing.Connect(otherThing);

			newThing.Compare(thing).should.be.false;
			thing.Compare(newThing).should.be.false;
		});

		it("should be different when the connections are different", ()=> {
			var newThing = new ThingModel.Thing(thing.ID, type);
			newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));

			newThing.Compare(thing).should.be.false;
			thing.Compare(newThing).should.be.false;
		});

		it("should handle differents ids correctly", ()=> {
			var newThing = new ThingModel.Thing("banana", type);
			newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));
			newThing.Connect(otherThing);

			newThing.Compare(thing, false, false).should.be.true;
			thing.Compare(newThing, false, false).should.be.true;

			thing.Compare(newThing, true, false).should.be.false;
		});

		describe("Properties comparisons", () => {


			it("should be different when the list of properties is different", () => {

				var newThing = new ThingModel.Thing(thing.ID, type);
				newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));
				newThing.Connect(otherThing);

				// New property
				newThing.SetProperty(new ThingModel.Property.String("surname", "Lapinou"));

				newThing.Compare(thing).should.be.false;
				thing.Compare(newThing).should.be.false;

				thing.SetProperty(new ThingModel.Property.String("surname", "Lapinou"));

				newThing.Compare(thing).should.be.true;
				thing.Compare(newThing).should.be.true;

				thing.SetProperty(new ThingModel.Property.String("color", "white"));
				newThing.SetProperty(new ThingModel.Property.String("meal", "carrot"));

				newThing.Compare(thing).should.be.false;
				thing.Compare(newThing).should.be.false;
			});

			it("should be different if the properties value are different", ()=> {

				var newThing = new ThingModel.Thing(thing.ID, type);
				newThing.SetProperty(new ThingModel.Property.String("name", "Bob"));
				newThing.Connect(otherThing);

				newThing.Compare(thing).should.be.false;
				thing.Compare(newThing).should.be.false;

				newThing.SetProperty(new ThingModel.Property.Double("name", 12));

				newThing.Compare(thing).should.be.false;
				thing.Compare(newThing).should.be.false;

				newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));

				newThing.Compare(thing).should.be.true;
				thing.Compare(newThing).should.be.true;
			});
		});

		it("should handle deep comparisons", ()=> {
			var aThingForTheRoad = new ThingModel.Thing("rabbit");
			aThingForTheRoad.SetProperty(new ThingModel.Property.Double("speed", 12));
			thing.Connect(aThingForTheRoad);

			var newThing = new ThingModel.Thing(thing.ID, type);
			newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));
			newThing.Connect(otherThing);
			newThing.Connect(aThingForTheRoad);

			thing.Compare(thing, true, true).should.be.true;
			thing.Compare(newThing, true, true).should.be.true;
			newThing.Compare(thing, true, true).should.be.true;

			newThing.Disconnect(otherThing);
			var newOtherThing = new ThingModel.Thing(otherThing.ID);
			newThing.Connect(newOtherThing);

			thing.Compare(newThing, true, true).should.be.true;
			newThing.Compare(thing, true, true).should.be.true;

			newOtherThing.SetProperty(new ThingModel.Property.String("name", "Alain"));

			thing.Compare(newThing, true, true).should.be.false;
			newThing.Compare(thing, true, true).should.be.false;
		});

		it("should handle infinite loop connections", ()=> {

			var newThing = new ThingModel.Thing(thing.ID, type);
			newThing.SetProperty(new ThingModel.Property.String("name", "Pierre"));
			newThing.Connect(otherThing);

			// infinite loop
			otherThing.Connect(thing);

			thing.Compare(newThing, true, true).should.be.true;
			newThing.Compare(thing, true, true).should.be.true;

			var newOtherThing = new ThingModel.Thing(otherThing.ID);
			newThing.Connect(newOtherThing);

			thing.Compare(newThing, true, true).should.be.false;
			newThing.Compare(thing, true, true).should.be.false;

			newOtherThing.Connect(thing);

			thing.Compare(newThing, true, true).should.be.true;
			newThing.Compare(thing, true, true).should.be.true;
		});
	});


// ReSharper restore WrongExpressionStatement
});